lib/web_ui/lib/src/engine/layers.dart

```diff
@@ -0,0 +1,750 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:typed_data';
+
+import 'package:meta/meta.dart';
+import 'package:ui/src/engine/scene_painting.dart';
+import 'package:ui/src/engine/vector_math.dart';
+import 'package:ui/ui.dart' as ui;
+
+class EngineRootLayer with PictureEngineLayer {}
+
+class BackdropFilterLayer
+  with PictureEngineLayer
+  implements ui.BackdropFilterEngineLayer {}
+class BackdropFilterOperation implements LayerOperation {
+  BackdropFilterOperation(this.filter, this.mode);
+
+  final ui.ImageFilter filter;
+  final ui.BlendMode mode;
+
+  @override
+  ui.Rect cullRect(ui.Rect contentRect) => contentRect;
+
+  @override
+  ui.Rect inverseMapRect(ui.Rect rect) => rect;
+
+  @override
+  void pre(SceneCanvas canvas, ui.Rect contentRect) {
+    canvas.saveLayerWithFilter(contentRect, ui.Paint()..blendMode = mode, filter);
+  }
+
+  @override
+  void post(SceneCanvas canvas, ui.Rect contentRect) {
+    canvas.restore();
+  }
+
+  @override
+  PlatformViewStyling createPlatformViewStyling() => const PlatformViewStyling();
+}
+
+class ClipPathLayer
+  with PictureEngineLayer
+  implements ui.ClipPathEngineLayer {}
+class ClipPathOperation implements LayerOperation {
+  ClipPathOperation(this.path, this.clip);
+
+  final ui.Path path;
+  final ui.Clip clip;
+
+  @override
+  ui.Rect cullRect(ui.Rect contentRect) => contentRect.intersect(path.getBounds());
+
+  @override
+  ui.Rect inverseMapRect(ui.Rect rect) => rect;
+
+  @override
+  void pre(SceneCanvas canvas, ui.Rect contentRect) {
+    canvas.save();
+    canvas.clipPath(path, doAntiAlias: clip != ui.Clip.hardEdge);
+    if (clip == ui.Clip.antiAliasWithSaveLayer) {
+      canvas.saveLayer(path.getBounds(), ui.Paint());
+    }
+  }
+
+  @override
+  void post(SceneCanvas canvas, ui.Rect contentRect) {
+    if (clip == ui.Clip.antiAliasWithSaveLayer) {
+      canvas.restore();
+    }
+    canvas.restore();
+  }
+
+  @override
+  PlatformViewStyling createPlatformViewStyling() {
```

Comment:
imo it would be cleaner to separate out the clip styling into a separate class. I don't think the layers should have this logic in them